### ISSUE TITLE
[Discussion] Permission system

### ISSUE THREAD

**Issue by Admicos** [okyanus-mc/api#9]
We need a permission system, how should we make it?

My implementation would be to create a Permission service interface, which all permission plugins should provide, and register to the server via something like the [Java service loader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html)

**Comment by portlek** [comment]
vault is also working like that, we can.
also the interface can be like that;
```java
public interface Permissible {
    /**
     * Whether or not the permissible object holds the given
     * permission string.
     *
     * @param perm the permission to check
     * @return {@code true} if the holder does have the
     * given permission
     */
    boolean hasPermission(String perm);

    /**
     * Adds the given permission to this holder's collection
     * of permissions.
     *
     * @param perm the permission to add
     */
    void addPermission(String perm);

    /**
     * Removes the permission from this holder's collection
     * of permissions.
     *
     * @param perm the permission to remove
     * @return {@code true} if the removal had the effect of
     * removing the permission, otherwise {@code false} if
     * it failed or it was not contained in the holder's
     * collection
     */
    boolean removePermission(String perm);

    /**
     * Sets whether or not the permissible is an operator,
     * or otherwise holds all permissions.
     *
     * @param op {@code true} to set to an operator,
     * {@code false} to set to a non-operator.
     */
    void setOp(boolean op);

    /**
     * Checks to see whether or not this permissible object
     * is set to an operator of the server.
     *
     * @return {@code true} if this permissible object is an
     * operator, {@code false} if it is not
     */
    boolean isOp();
}
```

**Comment by portlek** [comment]
for more example, we can look at vault permission service.

**Comment by Admicos** [comment]
I'm starting work on this, and pushed my initial implementation, any improvements?

**Comment by Admicos** [comment]
I also wrote a quick mod that implements this permission API, just to test things, and I didn't see any issues for now.

https://github.com/okyanus-mods/OKPerms

**Comment by Admicos** [comment]
As the initial implementation has shipped with 0.1.0, I feel safe closing this issue. If we need anything more, extending PermissionService *should* be easy.
